db/active_replicator_checkpointer.go

```diff
@@ -329,62 +355,87 @@ func (c *Checkpointer) fetchCheckpoints() error {
 	base.DebugfCtx(c.ctx, base.KeyReplicate, "got remote checkpoint: %v", remoteCheckpoint)
 	c.lastRemoteCheckpointRevID = remoteCheckpoint.Rev
 
-	// If checkpoint hash has changed, reset checkpoint to zero.  Shouldn't need to persist the updated checkpoints in
-	// the case both get rolled back to zero, can wait for next persistence
-	if localCheckpoint.ConfigHash != c.configHash {
-		localCheckpoint.LastSeq = ""
-	}
-
-	if remoteCheckpoint.ConfigHash != c.configHash {
-		remoteCheckpoint.LastSeq = ""
-	}
-
-	// If localSeq and remoteSeq match, we'll use this value.
 	localSeq := localCheckpoint.LastSeq
 	remoteSeq := remoteCheckpoint.LastSeq
+
+	// If localSeq and remoteSeq match, we'll use localSeq as the value.
 	checkpointSeq := localSeq
 
-	// Determine the lowest sequence if they don't match
-	if localSeq != remoteSeq {
-		base.DebugfCtx(c.ctx, base.KeyReplicate, "sequences mismatched, finding lowest of %q %q", localSeq, remoteSeq)
-		localSeqVal, err := parseIntegerSequenceID(localSeq)
-		if err != nil {
-			return err
+	// If both local and remote SGR2 checkpoints are missing, fetch SGR1 checkpoint as fallback as a one-time upgrade.
+	// Explicit checkpoint resets are expected to only reset one side of the checkpoints, so shouldn't fall into this handling.
+	var usedSGR1Checkpoint bool
+	if localSeq == "" && remoteSeq == "" && c.sgr1CheckpointID != "" {
+		base.Infof(base.KeyReplicate, "Attempting to fetch SGR1 checkpoint as fallback: %v", c.sgr1CheckpointID)
+		if c.sgr1CheckpointOnRemote {
+			base.DebugfCtx(c.ctx, base.KeyReplicate, "getting SGR1 checkpoint from remote (passive)")
+			checkpointSeq, _, err = c.getRemoteSGR1Checkpoint()
+		} else {
+			base.DebugfCtx(c.ctx, base.KeyReplicate, "getting SGR1 checkpoint from local (active)")
+			checkpointSeq, _, err = c.getLocalSGR1Checkpoint()
 		}
-
-		remoteSeqVal, err := parseIntegerSequenceID(remoteSeq)
-		if err != nil {
-			return err
+		if err != nil || checkpointSeq == "" {
+			c.stats.GetCheckpointSGR1FallbackMissCount++
+			base.Warnf("Unable to get SGR1 checkpoint, continuing without: %v", err)
+		} else {
+			base.DebugfCtx(c.ctx, base.KeyReplicate, "using upgraded sg-replicate checkpoint seq: %q", checkpointSeq)
+			usedSGR1Checkpoint = true
+			c.stats.GetCheckpointSGR1FallbackHitCount++
 		}
+	}
 
-		// roll local/remote checkpoint back to lowest of the two
-		if remoteSeqVal.Before(localSeqVal) {
-			checkpointSeq = remoteSeq
-			newLocalCheckpoint := remoteCheckpoint.Copy()
-			newLocalCheckpoint.Rev = c.lastLocalCheckpointRevID
-			c.lastLocalCheckpointRevID, err = c.setLocalCheckpoint(newLocalCheckpoint)
+	if !usedSGR1Checkpoint {
```

Comment:
Returning early in the usedSGR1Checkpoint case would make this easier to follow and keep the sgr1 stuff self-contained - I think you'd just need to increment stats and set lastCheckpointSeq?  

Alternately you could use the sgr1 result to build localSeq and remoteSeq (including configHash) and then just let the code below execute normally.  In either case I think that helps encapsulate the work that's sgr1-specific.